src/v/pandaproxy/reply.h

```diff
@@ -81,45 +82,73 @@ errored_body(std::error_code ec, ss::sstring msg) {
 }
 
-inline std::unique_ptr<ss::httpd::reply> unprocessable_entity(ss::sstring msg) {
-    return errored_body(
-      make_error_condition(reply_error_code::kafka_bad_request),
-      std::move(msg));
+inline std::unique_ptr<ss::httpd::reply>
+handle_errors(std::error_condition ec, ss::sstring msg, error_probe& eprobe) {
+    auto rep = errored_body(ec, std::move(msg));
+
+    using status_type = ss::httpd::reply::status_type;
+
+    // See seastar/http/reply.hh::reply::status_type for a list
+    // of supported error status codes.
+    if (rep->_status >= status_type{500}) {
+        eprobe.increment_5xx();
+    } else if (rep->_status >= status_type{400}) {
+        eprobe.increment_4xx();
+    } else if (rep->_status >= status_type{300}) {
+        eprobe.increment_3xx();
+    }
+
+    return rep;
 }
 
-inline std::unique_ptr<ss::httpd::reply> exception_reply(std::exception_ptr e) {
+inline std::unique_ptr<ss::httpd::reply>
+handle_errors(std::error_code ec, ss::sstring msg, error_probe& eprobe) {
+    return handle_errors(make_error_condition(ec), std::move(msg), eprobe);
+}
+
+inline std::unique_ptr<ss::httpd::reply>
+exception_reply(std::exception_ptr e, error_probe& eprobe) {
     try {
         std::rethrow_exception(e);
     } catch (const ss::gate_closed_exception& e) {
-        auto eb = errored_body(
-          reply_error_code::kafka_retriable_error, e.what());
+        auto eb = handle_errors(
+          reply_error_code::kafka_retriable_error, e.what(), eprobe);
         set_reply_unavailable(*eb);
         return eb;
     } catch (const json::exception_base& e) {
-        return errored_body(e.error, e.what());
+        return handle_errors(e.error, e.what(), eprobe);
     } catch (const parse::exception_base& e) {
-        return errored_body(e.error, e.what());
+        return handle_errors(e.error, e.what(), eprobe);
     } catch (const kafka::exception_base& e) {
-        return errored_body(e.error, e.what());
+        return handle_errors(e.error, e.what(), eprobe);
     } catch (const schema_registry::exception_base& e) {
-        return errored_body(e.code(), e.message());
+        return handle_errors(e.code(), e.message(), eprobe);
     } catch (const seastar::httpd::base_exception& e) {
-        return errored_body(
-          reply_error_code::kafka_bad_request,
-          e.what()); // TODO BP: Yarr!!
+        return handle_errors(
+          reply_error_code::kafka_bad_request, e.what(), eprobe);
     } catch (...) {
         vlog(plog.error, "{}", std::current_exception());
+        eprobe.increment_5xx();
         throw;
     }
 }
 
-struct exception_replier {
+class exception_replier {
+public:
     ss::sstring mime_type;
+
+    exception_replier(ss::sstring mt, error_probe& eprobe)
+      : mime_type(mt)
+      , _eprobe(eprobe) {}
+
     std::unique_ptr<ss::httpd::reply> operator()(const std::exception_ptr& e) {
-        auto res = exception_reply(e);
+        auto res = exception_reply(e, _eprobe);
```

Comment:
Rather than wrapping the `errored_body` with `handle_error in the `exception_reply`, why not just inspect the status code here?

I also don't see how we're counting errors that don't result from exceptions? There's probably the happy path where we also do `res->set_mime_type(mime_type)` like we do below.

Reply:
> Rather than wrapping the `errored_body` with `handle_error` in the `exception_reply`, why not just inspect the status code here?

Fair point.

> There's probably the happy path where we also do `res->set_mime_type(mime_type)` like we do below.

As far as I understand it, control passes through the exception_replier only when an exception occurs. So this operator overload is only called when there **is** and exception correct?

Reply:
> As far as I understand it, control passes through the exception_replier only when an exception occurs. So this operator overload is only called when there is an exception correct?

Correct. But it's also possible to set a non 2xx status without throwing, and those wouldn't be counted unless the "happy" (non-exception) path is also hooked.